### PR TITLE
support lift-splat-shoot view transformer

### PR DESCRIPTION
**Pull request by HuangJunJie2017**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support lift-splat-shoot view transformer

## Modification

Please briefly describe what modification is made in this PR.

+ mmdet3d/models/necks/view_transformer.py

Does the modification introduce changes that break the back-compatibility of the downstream repos? no
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation. 
will add bevdet afterward

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/necks/__init__.py [hunk 1]
```diff
@@ -4,7 +4,9 @@
 from .imvoxel_neck import OutdoorImVoxelNeck
 from .pointnet2_fp_neck import PointNetFPNeck
 from .second_fpn import SECONDFPN
+from .view_transformer import ViewTransformerLiftSplatShoot
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/__init__.py:7]: Can use LSS for LiftSplatShoot for simplicity

File: mmdet3d/models/necks/view_transformer.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+from mmcv.runner import BaseModule
+
+from ..builder import NECKS
+
+
+class QuickCumsum(torch.autograd.Function):
+    """Back propagation accelerated cumulative sum operator."""
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:10]: May need more detailed docstring for clear description

Comment (Tai-Wang) [reply mmdet3d/models/necks/view_transformer.py:10]: about the input variables

File: mmdet3d/models/necks/view_transformer.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        back[kept] -= 1
+
+        val = gradx[back]
+
+        return val, None, None
+
+
+@NECKS.register_module()
+class ViewTransformerLiftSplatShoot(BaseModule):
+    """Lift-Splat-Shoot view transformer for transform image-view feature into
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:42]: Can provide a reference for LSS here
Please note the format of including references in docstring. Can refer to other classes (such as other detectors).

File: mmdet3d/models/necks/view_transformer.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    bird-eye-view feature.
+
+    Args:
+        grid_config (dict(axis:list(float,float,float))): Config of grid alone
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:46]: No need for details of dict in the type hint
Can just explain it in the description

File: mmdet3d/models/necks/view_transformer.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+            each axis in format of (lower_bound, upper_bound, interval). axis
+            in {x,y,z,depth}.
+        input_size (tuple(int)): Size of input images in format of (height,
+            width).
+        in_channels (int): Channels of input feature.
+        tran_channels (int): Channels of transformed feature.
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:52]: out_channels or feat_channels?

File: mmdet3d/models/necks/view_transformer.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+        self.create_frustum()
+        self.tran_channels = tran_channels
+        self.depthnet = nn.Conv2d(
+            in_channels, self.D + self.tran_channels, kernel_size=1, padding=0)
+        self.accelerate = accelerate
+        self.max_voxel_points = max_voxel_points
+        self.initial_flag = True
+
+    def gen_grid_infos(self, x, y, z, **kwargs):
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:83]: Why use such a method to init the grid info

File: mmdet3d/models/necks/view_transformer.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        super(ViewTransformerLiftSplatShoot, self).__init__()
+        self.grid_config = grid_config
+        self.gen_grid_infos(**self.grid_config)
+
+        self.input_size = input_size
+        self.downsample = downsample
+
+        self.create_frustum()
+        self.tran_channels = tran_channels
+        self.depthnet = nn.Conv2d(
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:77]: self.depth_net

File: mmdet3d/models/necks/view_transformer.py (hunk 6 above)
Comment (Tai-Wang) [reply mmdet3d/models/necks/view_transformer.py:83]: Is there any big difference from initializing the grid with simple xxx.new_tensor/torch.tensor?

File: mmdet3d/models/necks/view_transformer.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    def get_lidar_coor(self, rots, trans, intrins, post_rots, post_trans):
+        """Calculate the locations of the frustum points in the lidar(ego)
+        coordinate system.
+
+        Args:
+            rots (torch.Tensor): of shape (N, N_cams, 3, 3). Rotation from
+                camera coordinate system to lidar coordinate system.
+            trans (torch.Tensor): of shape (N, N_cams, 3). Translation from
+                camera coordinate system to lidar coordinate system.
+            intrins (torch.Tensor): of shape (N, N_cams, 3, 3). Camera
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:134]: intrinsics

File: mmdet3d/models/necks/view_transformer.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+            .view(1, 1, W_feat).expand(self.D, H_feat, W_feat)
+        y = torch.linspace(0, H_in - 1, H_feat,  dtype=torch.float)\
+            .view(1, H_feat, 1).expand(self.D, H_feat, W_feat)
+
+        # D x H x W x 3
+        frustum = torch.stack((x, y, d), -1)
+        self.frustum = nn.Parameter(frustum, requires_grad=False)
+
+    def get_lidar_coor(self, rots, trans, intrins, post_rots, post_trans):
+        """Calculate the locations of the frustum points in the lidar(ego)
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:126]: lidar or ego? They can be different for different cases while the function name is get_lidar_coor. would be better to have a deterministic function

File: mmdet3d/models/necks/view_transformer.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+        points = torch.cat(
+            (points[..., :2, :] * points[..., 2:3, :], points[..., 2:3, :]), 5)
+        combine = rots.matmul(torch.inverse(intrins))
+        points = combine.view(B, N, 1, 1, 1, 3, 3).matmul(points).squeeze(-1)
+        points += trans.view(B, N, 1, 1, 1, 3)
+
+        return points
+
+    def voxel_pooling_prepare(self, coor, x):
+        """Data preparation for voxel pooling."""
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:161]: Please refine the docstring

File: mmdet3d/models/necks/view_transformer.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+        kept = (coor[:, 0] >= 0) & (coor[:, 0] < self.grid_size[0]) & \
+               (coor[:, 1] >= 0) & (coor[:, 1] < self.grid_size[1]) & \
+               (coor[:, 2] >= 0) & (coor[:, 2] < self.grid_size[2])
+        x, coor, point_idx = x[kept], coor[kept], point_idx[kept]
+
+        # get tensors from the same voxel next to each other
+        ranks = coor[:, 0] * (self.grid_size[1] * self.grid_size[2] * B)
+        ranks += coor[:, 1] * (self.grid_size[2] * B)
+        ranks += coor[:, 2] * B + coor[:, 3]
+        sorts = ranks.argsort()
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:190]: sorts -> order

File: mmdet3d/models/necks/view_transformer.py [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+
+    def voxel_pooling(self, coor, x):
+        """Generate bird-eye-view features with the pseudo point cloud.
+
+        Args:
+            coor (torch.tensor): of shape (B, N, D,  H, W, 3). Coordinate
+                of points.
+            x (torch.tensor): of shape (B, N, D,  H, W, C). Feature of points.
+
+        Returns: Bird-eye-view features of shape (B, C, H_BEV, W_BEV).
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:201]: Please correct the format

File: mmdet3d/models/necks/view_transformer.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+        ranks += coor[:, 1] * (self.grid_size[2] * B)
+        ranks += coor[:, 2] * B + coor[:, 3]
+        sorts = ranks.argsort()
+        return x[sorts], coor[sorts], ranks[sorts], point_idx[sorts]
+
+    def voxel_pooling(self, coor, x):
+        """Generate bird-eye-view features with the pseudo point cloud.
+
+        Args:
+            coor (torch.tensor): of shape (B, N, D,  H, W, 3). Coordinate
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:197]: Has one more space between D and H

File: mmdet3d/models/necks/view_transformer.py [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+        grid_size = self.grid_size.to(torch.long)
+        final = torch.zeros((B, C, grid_size[2], grid_size[1], grid_size[0]),
+                            device=x.device)
+        final[coor[:, 3], :, coor[:, 2], coor[:, 1], coor[:, 0]] = x
+        # collapse Z
+        final = torch.cat(final.unbind(dim=2), 1)
+
+        return final
+
+    def acceleration_initialize(self, coor, x):
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:219]: init_acceleration

File: mmdet3d/models/necks/view_transformer.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        """Conducting voxel pooling in accelerated mode.
+
+        Args:
+            x (torch.tensor): of shape (B, N, D, H, W, C). The feature of the
+                volumes.
+
+        Returns: Bird-eye-view features of shape (B, C, H_BEV, W_BEV).
+        """
+        B, N, D, H, W, C = x.shape
+        Nprime = B * N * D * H * W
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:263]: Nprime -> num_points?

File: tests/test_models/test_necks/test_necks.py [hunk 16]
```diff
@@ -132,3 +132,58 @@ def test_dla_neck():
...
+        input_size=(256, 704),
+        downsample=16,
+        in_channels=512,
+        tran_channels=64,
+        accelerate=False)
+    neck = build_neck(neck_cfg)
+    neck.init_weights()
+    neck = neck.cuda()
+
+    iv_feats = torch.rand(1, 2, 512, 44, 16).cuda()
```
Comment (Tai-Wang) [inline tests/test_models/test_necks/test_necks.py:156]: this variable name is a little confusing

File: mmdet3d/models/necks/view_transformer.py [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+    The function implementation process is as follows:
+
+        - step 1: Cumulatively sum the point-wise feature alone the point
+            queue.
+        - step 2: Remove the duplicated points with the same voxel rank and
+            only retain the last one in the point queue.
+        - step 3: Subtract each point feature with the previous one to obtain
+            the cumulative sum of the points in the same voxel.
+
+    Args:
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:24]: Please move the arguments below forward. The arguments docstring of a class typically corresponds to the args of the init function.

File: mmdet3d/models/necks/view_transformer.py [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+        x (torch.tensor): Point-wise features in shape (N_Points, C).
+        coor (torch.tensor): of shape (N_Points, D). The coordinate of points
+            in the feature coordinate system.
+        ranks (torch.tensor): of shape (N_Points). The rank of voxel that a
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:28]: "of shape" can be specified at the end of the description

File: mmdet3d/models/necks/view_transformer.py [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+
+        self.create_frustum(grid_config['depth'])
+        self.out_channels = out_channels
+        self.depth_net = nn.Conv2d(
+            in_channels, self.D + self.out_channels, kernel_size=1, padding=0)
+        self.accelerate = accelerate
+        self.max_voxel_points = max_voxel_points
+        self.initial_flag = True
+
+    def gen_grid_infos(self, x, y, z, **kwargs):
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:107]: would better be renamed as create_grid to be consistent with create_frustum?

File: mmdet3d/models/necks/view_transformer.py [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+            .view(-1, 1, 1).expand(-1, H_feat, W_feat)
+        self.D = d.shape[0]
+        x = torch.linspace(0, W_in - 1, W_feat,  dtype=torch.float)\
+            .view(1, 1, W_feat).expand(self.D, H_feat, W_feat)
+        y = torch.linspace(0, H_in - 1, H_feat,  dtype=torch.float)\
+            .view(1, H_feat, 1).expand(self.D, H_feat, W_feat)
+
+        # D x H x W x 3
+        frustum = torch.stack((x, y, d), -1)
+        self.frustum = nn.Parameter(frustum, requires_grad=False)
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:144]: Can also use torch.tensor for init?

File: mmdet3d/models/necks/view_transformer.py [hunk 21]
```diff
@@ -0,0 +1,347 @@
...
+
+    def voxel_pooling_prepare(self, coor, x):
+        """Data preparation for voxel pooling.
+
+        Args:
+            coor (torch.tensor): of shape (B, N, D, H, W, 3). Coordinate
+                of points in the lidar space.
+            x (torch.tensor): of shape (B, N, D, H, W, C). Feature of points.
+
+        Returns:
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:191]: tuple[torch.Tensor]: Feature of points of shape xxx; Coordinate of points in the voxel space of shape xxx;

File: mmdet3d/models/necks/view_transformer.py [hunk 22]
```diff
@@ -0,0 +1,347 @@
...
+                of points in lidar space.
+            x (torch.tensor): of shape (B, N, D, H, W, C). Feature of points.
+        """
+        x, coor, ranks, point_idx = self.voxel_pooling_prepare(coor, x)
+        # count for the repeat times of the same voxel rank in the point queue.
+        repeat_times = torch.ones(
+            coor.shape[0], device=coor.device, dtype=coor.dtype)
+        times = 0
+        repeat_times[0] = 0
+        curr_rank = ranks[0]
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/view_transformer.py:269]: Means the current rank? Can just use `cur_rank`

File: mmdet3d/models/necks/__init__.py [hunk 23]
```diff
@@ -4,7 +4,9 @@
 from .imvoxel_neck import OutdoorImVoxelNeck
 from .pointnet2_fp_neck import PointNetFPNeck
 from .second_fpn import SECONDFPN
+from .view_transformer import ViewTransformerLSS
 
 __all__ = [
-    'FPN', 'SECONDFPN', 'OutdoorImVoxelNeck', 'PointNetFPNeck', 'DLANeck'
+    'FPN', 'SECONDFPN', 'OutdoorImVoxelNeck', 'PointNetFPNeck', 'DLANeck',
+    'ViewTransformerLSS'
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/__init__.py:11]: LSSViewTransformer would be better?

**Review by Tai-Wang — COMMENTED** [review]
Just need to fix minor comments, especially for some naming conventions

File: mmdet3d/models/necks/view_transformer.py (hunk 19 above)
Comment (HuangJunJie2017) [reply mmdet3d/models/necks/view_transformer.py:107]: create_grid_infos? it is not for creating grids but the information of the grids instead.

**Review by Tai-Wang — APPROVED** [review]

**Pull request merged**